pkg/lock/simple_redis.go

```diff
@@ -64,45 +51,109 @@ type redisLock struct {
 	mu     sync.Mutex
 	key    string
 	token  string
-	log    *logrus.Entry
-	rng    *rand.Rand
+	// readers is the number of readers when the lock is acquired for reading
+	// or 0 when it is unlocked, or -1 when it is locked for writing.
+	readers int
+	log     *logrus.Entry
+	rng     *rand.Rand
 }
 
-func (rl *redisLock) Lock() error {
-	rl.mu.Lock()
-	defer rl.mu.Unlock()
+func (rl *redisLock) extends(writing bool) (bool, error) {
+	if rl.token == "" {
+		return false, nil
+	}
+	if (writing && rl.readers > 0) || (!writing && rl.readers < 0) {
+		return false, nil
```

Comment:
Shouldn't we return an error when trying to extend a lock we can't have since we're either getting a read lock on a write locked resource or a write lock on a read locked resource?

Not returning an error here, means we'll attempt to obtain a lock we can't obtain in `extendsOrObtainsReading()` for example.

Reply:
Well, with the current code, if we return an error, the `Lock()` or `RLock()` will fail with an error, without trying again later (when it should succeed). So, returning an error with the current code is not an option.

Currently, it tries to obtain the lock (even if we are sure that it will fail). We can optimize that but it's tricky. I prefer to keep the code simpler (even if it makes some useless calls to redis) instead of implementing the optimizations and taking the risk to introduce very subtle bugs in those optimizations.